llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -4362,6 +4379,25 @@ SDValue AArch64TargetLowering::LowerFP_TO_INT_SAT(SDValue Op,
   return DAG.getNode(ISD::TRUNCATE, DL, DstVT, Sat);
 }
 
+SDValue AArch64TargetLowering::LowerVectorXRINT(SDValue Op,
+                                                SelectionDAG &DAG) const {
+  EVT VT = Op.getValueType();
+  SDValue Src = Op.getOperand(0);
+  SDLoc DL(Op);
+
+  assert(VT.isVector() && "Expected vector type");
+
+  EVT CastVT =
+      VT.changeVectorElementType(Src.getValueType().getVectorElementType());
+
+  // Round the floating-point value into a floating-point register with the
+  // current rounding mode.
+  SDValue FOp = DAG.getNode(ISD::FRINT, DL, CastVT, Src);
+
+  // Truncate the rounded floating point to an integer, rounding to zero.
+  return DAG.getNode(ISD::FP_TO_SINT, DL, VT, FOp);
```

Comment:
I think you need to use ISD::FP_TO_SINT_SAT to represent a saturating conversion.  (It shouldn't affect the codegen on AArch64, but if you use the wrong node, other combines might reason about it incorrectly.)

Reply:
I think using FP_TO_SINT_SAT is wrong, as FP_TO_SINT_SAT returns zero on NaN values, while L[L]LRINT raise exceptions on NaN.

Reply:
ISD::LRINT can't raise an exception (at least, nothing can tell if it does); only ISD::STRICT_LRINT can, and you're not handling that here.

LangRef says the return value of an out-of-range LRINT is "equivalent to freeze poison".  But FP_TO_SINT just returns poison... so you can't use it.

I guess you could, alternatively, freeze the result of FP_TO_SINT, but that isn't really an improvement.